File: hashing/counting/c++/unique_count/unique_count.cpp
// unique_count.hpp
#include "unique_count.hpp"
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <algorithm>
#include <iterator>

bool uniqueCount(const std::vector<int> &nums)
{
    if (nums.size() == 1)
    {
        return true;
    }

    // O(n) to build
    std::unordered_map<int, int> counts;
    for (const auto &num : nums)
    {
        auto result = counts.try_emplace(num, 1);
        // If key already exists
        if (!result.second)
        {
            result.first->second++;
        }
    }

    // If all elements are unique with the same count, then this would cost O(n)
    std::unordered_set<int> setUniqueCounts;
    std::transform(
        // Iterate over the hash map
        counts.begin(), counts.end(),
        // An `std::insert_iterator` for inserting counts (at the end of the set)
        std::inserter(setUniqueCounts, setUniqueCounts.end()),
        // Lambda function to extract the count from each (num, count) pair
        [](const auto &pair)
        { return pair.second; });

    return (setUniqueCounts.size() == counts.size() ? true : false);
}

File: hashing/counting/python3/unique_count.py
from collections import Counter
from collections.abc import Sequence


def unique_count(nums: Sequence[int]) -> bool:
    """
    Given an array of integers, return true if all elements have unique counts.

    Parameters
    ----------
    nums : Sequence[int]
        A sequence of integers, which could be negative or positive

    Returns
    -------
    bool
        `True` if all elements have unique counts otherwise `False`
    """
    if len(nums) == 1:
        return True

    # O(n) to build
    counts = Counter(nums)

    # If all elements are unique with the same count, this would cost O(n)
    set_of_unique_counts = set(count for count in counts.values())

    return len(set_of_unique_counts) == len(counts)


def main() -> int:
    for nums in [[1, 2, 2, 1, 1, 3], [1, 2], [-3, 0, 1, -3, 1, 1, 1, -3, 10, 0]]:
        all_counts_unique = unique_count(nums)
        print(
            f"Given the input {nums}, {'all elements have unique counts' if all_counts_unique else 'not all elements have unique counts'}"
        )

    return 0


if __name__ == "__main__":
    main()
